File: benchmarks/kdtree-raycast/lang_cpp/triangle_mesh.cpp
#include "bounding_box.h"
#include "triangle_mesh.h"

TriangleMesh::TriangleMesh() {}

int32_t TriangleMesh::GetTrianglesCount() const
{
  return static_cast<int32_t>(triangles.size());
}

BoundingBox_f TriangleMesh::GetTriangleBounds(int32_t triangleIndex) const
{
  const auto& p = triangles[triangleIndex].points;
  auto bounds = BoundingBox_f(vertices[p[0].vertexIndex]);
  bounds.Extend(vertices[p[1].vertexIndex]);
  bounds.Extend(vertices[p[2].vertexIndex]);
  return bounds;
}

BoundingBox_f TriangleMesh::GetBounds() const
{
  BoundingBox_f bounds;
  for (int32_t i = 0; i < GetTrianglesCount(); i++) {
    bounds = BoundingBox_f::Union(bounds, GetTriangleBounds(i));
  }
  return bounds;
}

File: benchmarks/kdtree-construction/lang_go/triangle_mesh.go
package main

type TriangleMesh struct {
	vertices  []Vector32
	normals   []Vector32
	triangles [][3]int32
}

func (mesh *TriangleMesh) GetTrianglesCount() int32 {
	return int32(len(mesh.triangles))
}

func (mesh *TriangleMesh) GetTriangleBounds(triangleIndex int32) BBox32 {
	indices := mesh.triangles[triangleIndex]
	bbox := NewBBox32FromPoint(mesh.vertices[indices[0]])
	bbox.Extend(mesh.vertices[indices[1]])
	bbox.Extend(mesh.vertices[indices[2]])
	return bbox
}

func (mesh *TriangleMesh) GetBounds() BBox32 {
	meshBounds := NewBBox32()
	for i := 0; i < len(mesh.triangles); i++ {
		meshBounds = BBox32Union(meshBounds, mesh.GetTriangleBounds(int32(i)))
	}
	return meshBounds
}
